plugins/analysis/webinterface/recordedevents/recorded_events.go

```diff
@@ -1,94 +1,172 @@
+			if _, isAlready := links[incomingNeighborString]; !isAlready {
+				// Nope, so we have to allocate an empty map to be nested in links for incomingNeighborString
+				links[incomingNeighborString] = make(map[string]time.Time)
+			}
+
+			// Update graph when connection hasn't been seen before
+			if _, isAlready := links[incomingNeighborString][nodeIdString]; !isAlready {
+				server.Events.ConnectNodes.Trigger(incomingNeighborString, nodeIdString)
+			}
+			// Update links map
+			links[incomingNeighborString][nodeIdString] = timestamp
+		}
 	}))
+}
 
-	server.Events.ConnectNodes.Attach(events.NewClosure(func(sourceId string, targetId string) {
-		plugin.Node.Logger.Debugw("ConnectNodes", "sourceID", sourceId, "targetId", targetId)
-		lock.Lock()
-		defer lock.Unlock()
-
-		connectionMap, connectionMapExists := links[sourceId]
-		if !connectionMapExists {
-			connectionMap = make(map[string]bool)
-
-			links[sourceId] = connectionMap
+func Run() {
+	daemon.BackgroundWorker("Analysis Server Record Manager", func(shutdownSignal <-chan struct{}) {
+		ticker := time.NewTicker(CLEAN_UP_PERIOD)
+		for {
+			select {
+			case <-shutdownSignal:
+				return
+
+			case <-ticker.C:
+				cleanUp(CLEAN_UP_PERIOD)
+			}
 		}
-		connectionMap[targetId] = true
-	}))
+	}, shutdown.ShutdownPriorityAnalysis)
+}
 
-	server.Events.DisconnectNodes.Attach(events.NewClosure(func(sourceId string, targetId string) {
-		plugin.Node.Logger.Debugw("DisconnectNodes", "sourceID", sourceId, "targetId", targetId)
-		lock.Lock()
-		defer lock.Unlock()
+// Remove nodes and links we haven't seen for at least 3 times the heartbeat interval
+func cleanUp(interval time.Duration) {
+	lock.Lock()
+	defer lock.Unlock()
+	now := time.Now()
+
+	// Go through the list of connections. Remove connections that are older than interval time.
+	for srcNode, targetMap := range links {
+		for trgNode, lastSeen := range targetMap {
+			if now.Sub(lastSeen) > interval {
+				delete(targetMap, trgNode)
+				server.Events.DisconnectNodes.Trigger(srcNode, trgNode)
+			}
+		}
+		// Delete src node from links if it doesn't have any connections
+		if len(targetMap) == 0 {
+			delete(links, srcNode)
+		}
+	}
 
-		connectionMap, connectionMapExists := links[sourceId]
-		if connectionMapExists {
-			delete(connectionMap, targetId)
+	// Go through the list of nodes. Remove nodes that haven't been seen for interval time
+	for node, lastSeen := range nodes {
+		if now.Sub(lastSeen) > interval {
+			delete(nodes, node)
+			server.Events.NodeOffline.Trigger(node)
+			server.Events.RemoveNode.Trigger(node)
 		}
-	}))
+	}
 }
 
-func getEventsToReplay() (map[string]bool, map[string]map[string]bool) {
+func getEventsToReplay() (map[string]time.Time, map[string]map[string]time.Time) {
 	lock.Lock()
 	defer lock.Unlock()
 
-	copiedNodes := make(map[string]bool)
-	for nodeId, online := range nodes {
-		copiedNodes[nodeId] = online
+	copiedNodes := make(map[string]time.Time)
```

Comment:
```suggestion
	copiedNodes := make(map[string]time.Time, len(nodes))
```
might be preferable as the size of the map is known.